### ISSUE TITLE
first/second mutable borrow occurs here wrong diagnostic in nll

### ISSUE THREAD

**Issue by spastorino** [rust-lang/rust#57526]
After https://github.com/rust-lang/rust/pull/56113 we figured with @nikomatsakis that the output of `src/test/ui/borrowck/borrowck-mut-borrow-linear-errors.rs` in NLL is not ideal.

```rust
fn main() {
    let mut x = 1;
    let mut addr = vec![];
    loop {
        match 1 {
            1 => { addr.push(&mut x); } //[ast]~ ERROR [E0499]
            //[mir]~^ ERROR [E0499]
            2 => { addr.push(&mut x); } //[ast]~ ERROR [E0499]
            //[mir]~^ ERROR [E0499]
            _ => { addr.push(&mut x); } //[ast]~ ERROR [E0499]
            //[mir]~^ ERROR [E0499]
        }
    }
}
```

This outputs:

```
error[E0499]: cannot borrow `x` as mutable more than once at a time
  --> $DIR/borrowck-mut-borrow-linear-errors.rs:13:30
   |
LL |             1 => { addr.push(&mut x); } //[ast]~ ERROR [E0499]
   |                    ----      ^^^^^^ second mutable borrow occurs here
   |                    |
   |                    first borrow later used here
...
LL |             _ => { addr.push(&mut x); } //[ast]~ ERROR [E0499]
   |                              ------ first mutable borrow occurs here

error[E0499]: cannot borrow `x` as mutable more than once at a time
  --> $DIR/borrowck-mut-borrow-linear-errors.rs:15:30
   |
LL |             1 => { addr.push(&mut x); } //[ast]~ ERROR [E0499]
   |                    ---- first borrow later used here
LL |             //[mir]~^ ERROR [E0499]
LL |             2 => { addr.push(&mut x); } //[ast]~ ERROR [E0499]
   |                              ^^^^^^ second mutable borrow occurs here
LL |             //[mir]~^ ERROR [E0499]
LL |             _ => { addr.push(&mut x); } //[ast]~ ERROR [E0499]
   |                              ------ first mutable borrow occurs here

error[E0499]: cannot borrow `x` as mutable more than once at a time
  --> $DIR/borrowck-mut-borrow-linear-errors.rs:17:30
   |
LL |             _ => { addr.push(&mut x); } //[ast]~ ERROR [E0499]
   |                              ^^^^^^ mutable borrow starts here in previous iteration of loop

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0499`.
```

in NLL mode.

In particular the mention of first and second is weird. Why the `_ =>` arm is the first one?. Is that clear to the user?. There are several errors in the test also. Is it better to mix the fact that you're mutably borrowing something in `_ =>` and the mutably borrowing in `1 =>`? or is it better and easier to just state that both borrows are inside of the loop?.

**Comment by saleemjaffer** [comment]
@spastorino @estebank I would like to take a shot at this.

**Comment by spastorino** [comment]
@saleemjaffer go ahead!

**Comment by svmk** [comment]
I have the same problem. 
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=46df942353db38ad630798d7030c3177
It seems to me that my code should be valid. But for some reason, the borrow-checker considers the number of mutable references inside the method. Borrow-checker does not take into account the fact that the mutable link is used only once per method call.

**Comment by spastorino** [comment]
@svmk the problem you're facing is not the same that the one that is stated in this issue. In your case your mutable borrow last longer than what you think. If you need help to fix your code ask the question in https://users.rust-lang.org

**Comment by saleemjaffer** [comment]
> Is it better to mix the fact that you're mutably borrowing something in _ => and the mutably borrowing in 1 =>? or is it better and easier to just state that both borrows are inside of the loop?

@spastorino What is better to do out of the two approaches you have mentioned?

**Comment by estebank** [comment]
I think that stating that both are borrowing in a loop is the way to go, ideally highlighting the `loop` too.

**Comment by Aaron1011** [comment]
The error message is now the same both with and without `#![feature(nll)`.

**Comment by estebank** [comment]
Triage: the output has changed slightly, but it still doesn't point at the `loop`.

```
error[E0499]: cannot borrow `x` as mutable more than once at a time
  --> src/main.rs:6:30
   |
 6 |             1 => { addr.push(&mut x); }
   |                              ^^^^^^ second mutable borrow occurs here
...
10 |             _ => { addr.push(&mut x); }
   |                    ----      ------ first mutable borrow occurs here
   |                    |
   |                    first borrow later used here

error[E0499]: cannot borrow `x` as mutable more than once at a time
  --> src/main.rs:8:30
   |
 8 |             2 => { addr.push(&mut x); }
   |                              ^^^^^^ second mutable borrow occurs here
 9 |             
10 |             _ => { addr.push(&mut x); }
   |                    ----      ------ first mutable borrow occurs here
   |                    |
   |                    first borrow later used here

error[E0499]: cannot borrow `x` as mutable more than once at a time
  --> src/main.rs:10:30
   |
10 |             _ => { addr.push(&mut x); }
   |                    ----      ^^^^^^ `x` was mutably borrowed here in the previous iteration of the loop
   |                    |
   |                    first borrow used here, in later iteration of loop
```
